sycl/doc/Array_Kernel_Parameters.md

```diff
@@ -0,0 +1,435 @@
+For arrays, simply allowing them through would result in a
+function parameter of array type. This is not supported in C++.
+Therefore, the array needing capture is wrapped in a struct for
+the purposes of passing to the device. Once received on the device
+within its wrapper, the array is copied into the local capture object.
+All references to the array within the kernel body are directed to
+the non-wrapped array which is a member of the local capture object.
+
+<h4>Source code fragment:</h4>
+
+```C++
+  int array[100];
+  auto outBuf = buffer<int, 1>(&output[0], num_items);
+
+  myQueue.submit([&](handler &cgh) {
+    auto outAcc = outBuf.get_access<access::mode::write>(cgh);
+    cgh.parallel_for<class Worker>(num_items, [=](cl::sycl::id<1> index) {
+      outAcc[index] = array[index.get(0)];
+    });
+  });
+```
+
+<h4>Integration header produced:</h4>
+
+```C++
+static constexpr
+const kernel_param_desc_t kernel_signatures[] = {
+  //--- _ZTSZZ4mainENKUlRN2cl4sycl7handlerEE16->18clES2_E6Worker
+  { kernel_param_kind_t::kind_accessor, 4062, 0 },
+  { kernel_param_kind_t::kind_std_layout, 400, 32 },
+};
+```
+
+<h4>The changes to device code made to support this extension, in pseudo-code:</h4>
+
+```C++
+struct Capture {
+    Accessor outAcc;
+    int array[100];
+    () {
+        // Body
+    }
+}
+
+struct wrapper {
+    int array[100];
+};
+spir_kernel void caller(
+    __global int* AccData, // arg1 of Accessor init function
+    range<1> AccR1,        // arg2 of Accessor init function
+    range<1> AccR2,        // arg3 of Accessor init function
+    id<1> I,               // arg4 of Accessor init function
+    struct wrapper w_s     // Pass the array wrapped in a struct
+)
+{
+    // Local capture object
+    struct Capture local;
+
+    // Reassemble capture object from parts
+    // Initialize array using existing clang Initialization mechanisms
+    local.array = w_s; 
+    // Call accessor�s init function
+    Accessor::init(&local.outAcc, AccData, AccR1, AccR2, I);
+
+    callee(&local, id<1> wi);
+}
+```
+
+The sharp-eyed reviewer of `SemaSYCL.cpp` will notice that the array
+is actually double-wrapped in structs. This was done simply to preserve
+the interface to an existing function (`CreateAndAddPrmDsc`) which
+processes each kernel caller parameter as a capture object member. 
+The object being added to a list in `CreateAndAddPrmDsc` is `Fld`,
+which is expected to be a field of some struct. So a wrapped struct
+cannot be passed to this function. A double-wrapped struct is needed
+as shown below. This does not affect the generated code.
+
+```C++
+struct {
+  struct {
+    int array[100];
+  }
+}
+```
+
+This could be changed but it would mean changes to the `CreateAndAddPrmDsc`
+implementation, to all its callers and to the place where the list created
+by it is processed.
+By wrapping the array twice, the inner, single-wrapped array appears as a
+member of a struct and meets the requirements of the existing code.
```

Comment:
```suggestion
Single level of wrapping might look more elegant, but it would reduce potential for code reuse
and complicate the implementation w/o giving any benefits.
```

Reply:
this variant might be simpler